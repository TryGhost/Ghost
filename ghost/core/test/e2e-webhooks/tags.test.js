const {agentProvider, mockManager, fixtureManager, matchers} = require('../utils/e2e-framework');
const {anyGhostAgent, anyObjectId, anyISODateTime, anyString, anyContentVersion, anyNumber, anyLocalURL} = matchers;

const tagSnapshot = {
    created_at: anyISODateTime,
    description: anyString,
    id: anyObjectId,
    updated_at: anyISODateTime
};

describe('tag.* events', function () {
    let adminAPIAgent;
    let webhookMockReceiver;

    before(async function () {
        adminAPIAgent = await agentProvider.getAdminAPIAgent();
        await fixtureManager.init('integrations');
        await adminAPIAgent.loginAsOwner();
    });

    beforeEach(function () {
        webhookMockReceiver = mockManager.mockWebhookRequests();
    });

    afterEach(function () {
        mockManager.restore();
    });

    it('tag.added event is triggered', async function () {
        const webhookURL = 'https://test-webhook-receiver.com/tag-added/';
        await webhookMockReceiver.mock(webhookURL);
        await fixtureManager.insertWebhook({
            event: 'tag.added',
            url: webhookURL
        });

        await adminAPIAgent
            .post('tags/')
            .body({
                tags: [{
                    name: 'Test Tag',
                    slug: 'test-tag',
                    description: 'Test Description'
                }]
            })
            .expectStatus(201);

        await webhookMockReceiver.receivedRequest();

        webhookMockReceiver
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                'content-length': anyNumber,
                'user-agent': anyGhostAgent
            })
            .matchBodySnapshot({
                tag: {
                    current: {...tagSnapshot, url: anyLocalURL}
                }
            });
    });

    it('tag.deleted event is triggered', async function () {
        const webhookURL = 'https://test-webhook-receiver.com/tag-deleted/';
        await webhookMockReceiver.mock(webhookURL);
        await fixtureManager.insertWebhook({
            event: 'tag.deleted',
            url: webhookURL
        });

        const res = await adminAPIAgent
            .post('tags/')
            .body({
                tags: [{
                    name: 'Test Tag 2',
                    slug: 'test-tag-2',
                    description: 'Test description'
                }]
            })
            .expectStatus(201);

        const id = res.body.tags[0].id;

        await adminAPIAgent
            .delete('tags/' + id)
            .expectStatus(204);

        await webhookMockReceiver.receivedRequest();

        webhookMockReceiver
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                'content-length': anyNumber,
                'user-agent': anyGhostAgent
            })
            .matchBodySnapshot({
                tag: {
                    current: {},
                    previous: tagSnapshot
                }
            });
    });

    it('tag.edited event is triggered', async function () {
        const webhookURL = 'https://test-webhook-receiver.com/tag-edited/';
        await webhookMockReceiver.mock(webhookURL);
        await fixtureManager.insertWebhook({
            event: 'tag.edited',
            url: webhookURL
        });
    
        const res = await adminAPIAgent
            .post('tags/')
            .body({
                tags: [{
                    name: 'Test Tag 3',
                    slug: 'test-tag-3',
                    description: 'Test description'
                }]
            })
            .expectStatus(201);
    
        const id = res.body.tags[0].id;
    
        const updatedTag = res.body.tags[0];
        updatedTag.name = 'Updated Tag 3';
        updatedTag.slug = 'updated-tag-3';
    
        await adminAPIAgent
            .put('tags/' + id)
            .body({
                tags: [updatedTag]
            })
            .expectStatus(200);
    
        await webhookMockReceiver.receivedRequest();
    
        webhookMockReceiver
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                'content-length': anyNumber,
                'user-agent': anyGhostAgent
            })
            .matchBodySnapshot({
                tag: {
                    current: {...tagSnapshot, url: anyLocalURL},
                    previous: {updated_at: anyISODateTime}
                }
            });
    });
});
