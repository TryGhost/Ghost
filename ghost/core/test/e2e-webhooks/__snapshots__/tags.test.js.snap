// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tag.* events tag.added event is triggered 1: [headers] 1`] = `
Object {
  "accept-encoding": "gzip, deflate, br",
  "content-length": Any<Number>,
  "content-type": "application/json",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "user-agent": StringMatching /Ghost\\\\/\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\\\s\\\\\\(https:\\\\/\\\\/github\\.com\\\\/TryGhost\\\\/Ghost\\\\\\)/,
}
`;

exports[`tag.* events tag.added event is triggered 2: [body] 1`] = `
Object {
  "tag": Object {
    "current": Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": Any<String>,
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "Test Tag",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "test-tag",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": StringMatching /http:\\\\/\\\\/127\\.0\\.0\\.1:2369\\\\/\\[A-Za-z0-9_-\\]\\+\\\\//,
      "visibility": "public",
    },
    "previous": Object {},
  },
}
`;

exports[`tag.* events tag.deleted event is triggered 1: [headers] 1`] = `
Object {
  "accept-encoding": "gzip, deflate, br",
  "content-length": Any<Number>,
  "content-type": "application/json",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "user-agent": StringMatching /Ghost\\\\/\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\\\s\\\\\\(https:\\\\/\\\\/github\\.com\\\\/TryGhost\\\\/Ghost\\\\\\)/,
}
`;

exports[`tag.* events tag.deleted event is triggered 2: [body] 1`] = `
Object {
  "tag": Object {
    "current": Object {},
    "previous": Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": Any<String>,
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "Test Tag 2",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "test-tag-2",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "visibility": "public",
    },
  },
}
`;

exports[`tag.* events tag.edited event is triggered 1: [headers] 1`] = `
Object {
  "accept-encoding": "gzip, deflate, br",
  "content-length": Any<Number>,
  "content-type": "application/json",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "user-agent": StringMatching /Ghost\\\\/\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\\\s\\\\\\(https:\\\\/\\\\/github\\.com\\\\/TryGhost\\\\/Ghost\\\\\\)/,
}
`;

exports[`tag.* events tag.edited event is triggered 2: [body] 1`] = `
Object {
  "tag": Object {
    "current": Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": Any<String>,
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "Updated Tag 3",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "updated-tag-3",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": StringMatching /http:\\\\/\\\\/127\\.0\\.0\\.1:2369\\\\/\\[A-Za-z0-9_-\\]\\+\\\\//,
      "visibility": "public",
    },
    "previous": Object {
      "name": "Test Tag 3",
      "slug": "test-tag-3",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
    },
  },
}
`;
